File: main/src/main/java/com/sedmelluq/discord/lavaplayer/container/matroska/MatroskaAudioTrack.java
package com.sedmelluq.discord.lavaplayer.container.matroska;

import com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaFileTrack;
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import com.sedmelluq.discord.lavaplayer.track.BaseAudioTrack;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioProcessingContext;
import com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

/**
 * Audio track that handles the processing of MKV and WEBM formats
 */
public class MatroskaAudioTrack extends BaseAudioTrack {
  private static final Logger log = LoggerFactory.getLogger(MatroskaAudioTrack.class);

  private final SeekableInputStream inputStream;

  /**
   * @param trackInfo Track info
   * @param inputStream Input stream for the file
   */
  public MatroskaAudioTrack(AudioTrackInfo trackInfo, SeekableInputStream inputStream) {
    super(trackInfo);

    this.inputStream = inputStream;
  }

  @Override
  public void process(LocalAudioTrackExecutor localExecutor) {
    MatroskaStreamingFile file = loadMatroskaFile();
    MatroskaTrackConsumer trackConsumer = loadAudioTrack(file, localExecutor.getProcessingContext());

    try {
      localExecutor.executeProcessingLoop(() -> {
        file.provideFrames(trackConsumer);
      }, position -> {
        file.seekToTimecode(trackConsumer.getTrack().index, position);
      });
    } finally {
      trackConsumer.close();
    }
  }

  private MatroskaStreamingFile loadMatroskaFile() {
    try {
      MatroskaStreamingFile file = new MatroskaStreamingFile(inputStream);
      file.readFile();

      accurateDuration.set((int) file.getDuration());
      return file;
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private MatroskaTrackConsumer loadAudioTrack(MatroskaStreamingFile file, AudioProcessingContext context) {
    MatroskaTrackConsumer trackConsumer = null;
    boolean success = false;

    try {
      trackConsumer = selectAudioTrack(file.getTrackList(), context);

      if (trackConsumer == null) {
        throw new IllegalStateException("No supported audio tracks in the file.");
      } else {
        log.debug("Starting to play track with codec {}", trackConsumer.getTrack().codecId);
      }

      trackConsumer.initialise();
      success = true;
    } finally {
      if (!success && trackConsumer != null) {
        trackConsumer.close();
      }
    }

    return trackConsumer;
  }

  private MatroskaTrackConsumer selectAudioTrack(MatroskaFileTrack[] tracks, AudioProcessingContext context) {
    MatroskaTrackConsumer trackConsumer = null;

    for (MatroskaFileTrack track : tracks) {
      if (track.type == MatroskaFileTrack.Type.AUDIO) {
        if (MatroskaContainerProbe.OPUS_CODEC.equals(track.codecId)) {
          trackConsumer = new MatroskaOpusTrackConsumer(context, track);
          break;
        } else if (MatroskaContainerProbe.VORBIS_CODEC.equals(track.codecId)) {
          trackConsumer = new MatroskaVorbisTrackConsumer(context, track);
        } else if (MatroskaContainerProbe.AAC_CODEC.equals(track.codecId)) {
          trackConsumer = new MatroskaAacTrackConsumer(context, track);
        }
      }
    }

    return trackConsumer;
  }
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/container/matroska/MatroskaAudioTrack.kt
package com.sedmelluq.discord.lavaplayer.container.matroska

import com.sedmelluq.discord.lavaplayer.container.matroska.format.MatroskaFileTrack
import com.sedmelluq.discord.lavaplayer.tools.ExceptionTools
import com.sedmelluq.discord.lavaplayer.tools.io.SeekableInputStream
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo
import com.sedmelluq.discord.lavaplayer.track.BaseAudioTrack
import com.sedmelluq.discord.lavaplayer.track.playback.AudioProcessingContext
import com.sedmelluq.discord.lavaplayer.track.playback.LocalAudioTrackExecutor
import kotlinx.atomicfu.update
import mu.KotlinLogging
import java.io.IOException

/**
 * Audio track that handles the processing of MKV and WEBM formats
 *
 * @param trackInfo   Track info
 * @param inputStream Input stream for the file
 */
class MatroskaAudioTrack(trackInfo: AudioTrackInfo, private val inputStream: SeekableInputStream) : BaseAudioTrack(trackInfo) {
    companion object {
        private val log = KotlinLogging.logger { }
    }

    override fun process(executor: LocalAudioTrackExecutor) {
        val file = loadMatroskaFile()
        val trackConsumer = loadAudioTrack(file, executor.processingContext)
        try {
            executor.executeProcessingLoop(
                { file.provideFrames(trackConsumer) },
                { file.seekToTimecode(trackConsumer!!.track.index, it) }
            )
        } finally {
            ExceptionTools.closeWithWarnings(trackConsumer)
        }
    }

    private fun loadMatroskaFile(): MatroskaStreamingFile {
        return try {
            val file = MatroskaStreamingFile(inputStream)
            file.readFile()
            accurateDuration = file.duration.toLong()
            file
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private fun loadAudioTrack(file: MatroskaStreamingFile, context: AudioProcessingContext): MatroskaTrackConsumer? {
        var trackConsumer: MatroskaTrackConsumer? = null
        var success = false

        try {
            trackConsumer = selectAudioTrack(file.trackList, context)
            checkNotNull(trackConsumer) { "No supported audio tracks in the file." }
            log.debug("Starting to play track with codec {}", trackConsumer.track.codecId)
            trackConsumer.initialise()
            success = true
        } finally {
            if (!success && trackConsumer != null) {
                ExceptionTools.closeWithWarnings(trackConsumer)
            }
        }

        return trackConsumer
    }

    private fun selectAudioTrack(tracks: Array<MatroskaFileTrack>, context: AudioProcessingContext): MatroskaTrackConsumer? {
        var trackConsumer: MatroskaTrackConsumer? = null
        for (track in tracks) {
            if (track.type == MatroskaFileTrack.Type.AUDIO) {
                if (MatroskaContainerProbe.OPUS_CODEC == track.codecId) {
                    trackConsumer = MatroskaOpusTrackConsumer(context, track)
                    break
                } else if (MatroskaContainerProbe.VORBIS_CODEC == track.codecId) {
                    trackConsumer = MatroskaVorbisTrackConsumer(context, track)
                } else if (MatroskaContainerProbe.AAC_CODEC == track.codecId) {
                    trackConsumer = MatroskaAacTrackConsumer(context, track)
                }
            }
        }

        return trackConsumer
    }
}
